transport/http/src/main/java/org/kaazing/gateway/transport/http/PersistentConnectionPool.java

```diff
@@ -21,62 +21,90 @@
 
     private static final String IDLE_FILTER = HttpProtocol.NAME + "#idle";
+    private static final TypedAttributeKey<HttpResourceAddress> SERVER_ADDRESS =
+            new TypedAttributeKey<>(PersistentConnectionPool.class, "address");
+
+    // server address -> set of persistent connections (per thread).
+    // Using ThreadLocal for the following reasons:
+    // - we need to return thread-aligned IoSession
+    // - it reduces contention compared to synchronized ServerConnections
+    private final ThreadLocal<ServerConnections> connections;
+
+    // No of connections per server across ALL threads
+    private final Map<HttpResourceAddress, AtomicInteger> counts;
+
 
-    // transport address -> set of persistent connections
-    private final Map<ResourceAddress, Set<IoSession>> connections;
     private final Logger logger;
-    private final CloseListener closeListener;
     private final HttpConnectIdleFilter idleFilter;
+    private final CloseListener closeListener;
 
     PersistentConnectionPool(Logger logger) {
-        this.connections = new HashMap<>();
+        this.connections = new VicariousThreadLocal<ServerConnections>() {
+            @Override
+            protected  ServerConnections initialValue() {
+                return new ServerConnections();
+            }
+        };
+        this.counts = new ConcurrentHashMap<>();
         this.logger = logger;
+        this.idleFilter = new HttpConnectIdleFilter(logger);
         this.closeListener = new CloseListener(this);
-        this.idleFilter = new HttpConnectIdleFilter(this, logger);
     }
 
     /*
-     * Recycle existing transport session so that it can be used as a http
-     * persistent connection
+     * Cache existing transport session so that it can be used for future
+     * connect requests to server
+     *
+     * @return true if the idle connection is cached for reuse
+     *         false otherwise
      */
-    public void recycle(IoSession session, Integer keepAliveTimeout) {
-        assert keepAliveTimeout != null;
-        if (logger.isDebugEnabled()) {
-            logger.debug(String.format("Recycling (adding to pool) http connect persistent connection %s", session));
+    boolean recycle(DefaultHttpSession httpSession) {
+        if (!add(httpSession)) {
+            return false;
         }
-        ResourceAddress address = BridgeSession.REMOTE_ADDRESS.get(session);
-        Set<IoSession> sessions = connections.get(address);
-        if (sessions == null) {
-            sessions = new HashSet<>();
-            connections.put(address, sessions);
-        }
-        sessions.add(session);
-        CloseFuture closeFuture = session.getCloseFuture();
+
+        HttpResourceAddress serverAddress = (HttpResourceAddress)httpSession.getRemoteAddress();
+        IoSession transportSession = httpSession.getParent();
+
+        SERVER_ADDRESS.set(transportSession, serverAddress);
+
+        // Take care of transport session close
```

Comment:
What does "take care of" mean in this case?  Is this about ensuring the session is removed from the connection pool on transport close?